keps/sig-api-machinery/00xx-publish-crd-openapi.md

```diff
@@ -0,0 +1,142 @@
+---
+
+# Publish CRD OpenAPI
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+In CustomResourceDefinition (CRD) we allow CRD author to define JSON-Schema, to
+enable server-side [validation for CustomResources (CR)](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md).
+The JSON-Schema format is compatible for creating OpenAPI documentation for CRs,
+which can be used by clients like kubectl to perform client-side validation,
+schema explanation (`kubectl explain`), and client generation.
+This KEP purposes using the JSON-Schema to create and publish OpenAPI
+documentation for CRs.
+
+## Motivation
+
+Publishing CRD OpenAPI enables client-side validation, schema explanation and
+client generation for CRs. It covers the gap between CRD and native Kubernetes
+APIs, which already support OpenAPI documentation.
+
+Publishing CRD OpenAPI is also noted as potential followup in [Validation for CustomResources Implementation Plan](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md#implementation-plan).
+
+### Goals
+
+* For every CRD that we serve, publish Paths (operations that we support on
+  resource and subresources) and Definitions (for both CR object and CR list
+  object) in OpenAPI documentation to fully demonstrate the existence of the
+  API.
+* For CRDs with validation JSON-Schema defined, the CR object Definition should
+  include both CRD validation JSON-Schema and native Kubernetes ObjectMeta and
+  TypeMeta properties.
+* For CRDs without validation JSON-Schema, the CR object Definition will be an
+  emtpy object which validate arbitrary request on client-side.
+
+### Non-Goals
+
+* Support/honor OpenAPI AdditionalProperties in kube-openapi and kubectl, to
+  enable client-side validation on native Kubernetes ObjectMeta and TypeMeta
+  while allowing arbitrary extra fields in a request.
+* Finalize the design on how we aggregate OpenAPI documentation from
+  apiextensions-apiserver
+  - [Short term design doc](https://docs.google.com/document/d/13lBj8Stdwku8BgL0fbT__4Iw97NRh77loJ_MoZuCGwQ/edit#)
+    including multiple workarounds has been purposed and is under discussion.
+  - Long term design doc including generalizing APIService API or implementing
+    watch mechanism for /openapi/v2 endpoint on apiextension-apiserver is work
+    in progress.
+
+## Proposal
+
+ref: https://github.com/kubernetes/kubernetes/pull/71192
+
+### Risks and Mitigations
+
+TODO(roycaihw):
```

Comment:
simplify to something like this:

Performance: OpenAPI aggregation has performed poorly in the past. Introducing aggregation that occurs on a higher frequency could cause performance issues in the apiserver. To mitigate this, improvements in OpenAPI merging are being made (both in terms of memory and CPU cost), and benchmarking will be done prior to merge to ensure we do not regress API server performance.

Compatibility: some Kubernetes components (notably kubectl) perform client-side validation based on published openapi schemas. To ensure that publishing schemas for additional custom resources does not regress existing function, we will test creation of custom resources with and without validation with current and prior versions of kubectl.